text/3470-crabi-v1.md

```diff
@@ -0,0 +1,661 @@
+```
+
+is equivalent to:
+
+```c
+struct u32_boxed_slice {
+    uint32_t *data;
+    size_t len;
+};
+struct f64_boxed_slice {
+    double *data;
+    size_t len;
+};
+extern struct f64_boxed_slice func(struct u32_boxed_slice data);
+```
+
+## `Box<str>` - boxed counted UTF-8 string
+
+crABI supports passing an owned string. To represent this via the C ABI, crABI
+passes this equivalently to a boxed slice of type `Box<[u8]>`.
+
+## `[T; N]` - fixed-size array by value
+
+crABI supports passing a fixed-size array by value (as opposed to by
+reference). To represent this via the C ABI, crABI treats this equivalently to
+a C array of the same type passed by value.
+
+Note that this means C code can use an array directly in a context where it
+will be interpreted by-value (such as in a struct field), but needs to use a
+structure with a single array field in contexts where it would otherwise be
+interpreted as a pointer (such as in a function argument or return value).
+
+For instance:
+
+```rust
+extern "crabi" fn func(rgb: [u16; 3])
+```
+
+is equivalent to:
+
+```c
+struct func_rgb_arg {
+    uint16_t array[3];
+};
+extern void func(struct func_rgb_arg rgb);
+```
+
+Note that crABI does *not* pass the length, since it's a compile-time constant;
+the recipient must also know the correct size. (Use one of the slice-based
+types for a type with a runtime-determined length.)
+
+## `&[T; N]` - fixed-size array by reference
+
+crABI supports passing a fixed-size array by reference. crABI represents this
+as a pointer to the element type.
+
+C can represent this as an array (e.g. `uint16_t rgb[3]`) in contexts where
+that already implies passing by pointer (such as a function argument or return
+value), or translate it explicitly to a pointer (e.g `uint16_t (*rgb)[3]`) in
+contexts where just writing the array would imply by-value (such as a struct
+field).
+
+## Arbitrary `enum` types
+
+crABI supports arbitrary `enum` types, if declared with `repr(crabi)`. These
+are always passed using the same layout that Rust uses for enums with `repr(C)`
+and a specified discriminant type:
+<https://doc.rust-lang.org/reference/type-layout.html#combining-primitive-representations-of-enums-with-fields-and-reprc>
+
+This layout consists of the discriminant, followed by a union of the
+representations of each `enum` variant that has fields.
+
+If an `enum` specifies `repr(crabi)` but does not specify a discriminant type,
+the `enum` is guaranteed to use the smallest discriminant type that holds the
+maximum discriminant value used by a variant in the `enum`.
+
+If the `enum` has no fields, or no fields with a non-zero size, crABI will
+represent the `enum` as only its discriminant.
+
+### Guaranteed niche optimization
+[niche]: #niche
+
+As a special case, if an `enum` using `repr(crabi)` has exactly two variants,
+one of which has no fields and the other of which has a single field, and the
+single field type has a specific type (defined below) with a "niche" value,
```

Comment:
What about `non_exhaustive`?

Reply:
Is any `non_exhaustive` type eligible for a stable ABI like this? Or would it somehow be limited to extensions in ways that don't break the ABI? Then again, do we have restrictions on adding fields or other layout-affecting changes being made to extant ABI-declared types?

Reply:
`#[non_exhaustive]` doesn't have any effect for the crate that defines the type. Instead it has an effect on users. This means that if the user tried to use a newer version of the type, doing so while using the older version of the code that defined is would be UB without said code being able to report an error for the invalid discriminant. Also extensions to `#[non_exhaustive]` types can only possibly work for the memory layout. The calling convention can drastically change even for adding a single field or enum variant.